### DISCUSSION TITLE
Custom hyphenation (.pattern files) in Android and Linux apps

### DISCUSSION THREAD

**Discussion by PrasiNV** [koreader/koreader#12172] [Q&A] [answered]
I would like to know if it is possible to add custom hyphenation files (.pattern files) in the Android app and Linux appimage?
If yes, to which folder they have to be added?
I know how to do this on a Kobo. But unable to achieve the same in Android app and Linux appimage. Help please.
Thanks

**Comment by poire-z** [comment]
Pinging @zwim, who may be doing that. I think a 0- or 1- userpatch may be able to copy your file into the Android private koreader directories. cf https://github.com/koreader/koreader/wiki/User-patches .
Dunno about Linux/AppImage.

**Comment by ghost** [comment]
I have a patch for my personal usage with German. You can use it as a starting point. 
(If I recall correctly, in my pattern files there is the creation date. This is used for checking if your pattern has to be copied ...)


[10-UpdateGermanPattern.lua.txt](https://github.com/user-attachments/files/16264275/10-UpdateGermanPattern.lua.txt)

**Reply by PrasiNV** [reply]
Thank you for sharing the reference. I am not a programmer and need some help. I have a Android device with built-in storage (not an SD card). If I am not wrong, the location of the custom pattern file is `/storage/emulated/0/hyphenations/English_US.pattern`

My custom pattern file is a modified version of the original `English_US.pattern` file bundled with KOReader. As the original bundled pattern file do not have the date in the same format as the German.pattern file has it, I am not sure whether what I am doing is correct because after placing the attached patch file and the custom pattern file in corresponding places, I am unable to get the KOReader app copy the custom pattern file into original pattern file.

May I please ask you to point me where I am doing wrong? Thank you.

[10-UpdateEnglish_USPattern.lua.txt](https://github.com/user-attachments/files/16315262/10-UpdateEnglish_USPattern.lua.txt)


![English_US pattern](https://github.com/user-attachments/assets/d0fba84d-11bd-4866-b024-93a340a81f2d)

**Comment by ghost** [comment]
Sorry, I am short at time right for tje weekend. I will come back on monday

**Reply by PrasiNV** [reply]
A gentle reminder requesting your help please. Thank you.

**Comment by ghost** [comment, marked as answer]
I have played a bit now on my mobile. You can use

[10-UpdateEnglish_USPattern_Android.lua.txt](https://github.com/user-attachments/files/16377763/10-UpdateEnglish_USPattern_Android.lua.txt)

Important: place your new pattern in `/storage/emulated/0/koreader/hyphenations/English_US.pattern` .

**Reply by PrasiNV** [reply]
Thank you for coming back. I tried using the latest patch you provided and also placed the custom pattern in the location as you advised. Everything worked. Thanks again. I appreciate your help.

**Reply by ghost** [reply]
Thanks for your reply. I want/will simplify my amd your userpatch. I will ask you to test it in a few days.

**Reply by PrasiNV** [reply]
Sure. Please let me know. 

By the way, I have also tried to simplify it as attached by removing the date comparison checks and at every startup the copy of pattern file happens without any checks. It may sound a bit too much, but I tried and it too works without causing any issue.

[1-UpdateEnglish_USPattern.lua.txt](https://github.com/user-attachments/files/16381817/1-UpdateEnglish_USPattern.lua.txt)

**Reply by ghost** [reply]
at thefirst glance your patch should work well (for quite small paterns).

**Reply by ghost** [reply]
And I forgot to mention: if you have better patterns as our shipped ones, you might want to make a PR for that.
If you need help for doing that, please feel free to ask for help..

**Reply by PrasiNV** [reply]
I wish I could contribute to this wonderful software and helpful opensource community. The custom pattern I made for myself is something which is a combination of hyphenation patterns for English and three other Indian languages (each of which has their own script and unicode range).

The reason to combine multiple languages into one file is, I have collection of epubs which are not well formatted i.e. individual languages are not well structured to start with by having appropriate language tags. These epubs contain texts in combination of these four languages. To have proper rendering of these epubs I started with combining all patterns into the default English_US pattern and it seems to be meeting my specific requirement. As the pattern file is modified to suit a very specific requirement, I am not sure if this could be of help to wider community as the ask here is to combine other languages into default English_US pattern file. However, if someone having similar requirement asks me I am happy to share what is working for me.

**Reply by poire-z** [reply]
No, keep that niche/user solution for you, or share your hyph files here or in a github, or just wait for users to request it from you if they need it.

I didn't know Indic could be hyphenated. We don't have any hyph pattern for Indian languages/scripts. I'm not even sure our hyphenation code works with non-latin/cyrillic/greek scripts: does it really work ?!

In any case, we shouldn't really have shipped merged hyph dicts (we have it for Russian+English for historical reasons, crengine has it when we picked it as a subprokect) as there's so many combinations possible that we would end up with tons of combinations  (ie. why not have English+Russian+Greek+Bengali :))
May be one day, we'll be able to select multiple hyph dicts, I talked about that in https://github.com/koreader/crengine/pull/383#issuecomment-753472141 - but lots of work for niche use cases.

**Reply by PrasiNV** [reply]
Indic languages can be hyphenated (at least some of them as I learnt it recently). 

But I do not think they have exhaustive dictionaries as other languages have - at least I haven't came across. 

I started with these three resources:

[https://github.com/santhoshtr/hyphenation](https://github.com/santhoshtr/hyphenation)
[https://thottingal.in/blog/2022/02/18/indian-language-hyphenation/](https://thottingal.in/blog/2022/02/18/indian-language-hyphenation/)
[https://www.w3.org/TR/ilreq/#h_hyphenation](https://www.w3.org/TR/ilreq/#h_hyphenation)

Based on the above and some more research, I came across hyphenation files in tex format from below location and I have converted them into `.pattern` format

[https://hyphenation.org/](https://hyphenation.org/)

Finally I took the patterns from individual Indian language pattern files and placed them at the bottom of the English_US.pattern file which ships with KOReader. Here are the two pattern files which seem to be meeting my own requirements:

[English_US.pattern.txt](https://github.com/user-attachments/files/16393255/English_US.pattern.txt)
[English_GB.pattern.txt](https://github.com/user-attachments/files/16393257/English_GB.pattern.txt)

**Reply by poire-z** [reply]
Just for curiosity, can you post some screenshots of our rendering of text with some of its words hyphenated for the various Indic scripts you read?

**Reply by PrasiNV** [reply]
Image-1 contains Devanagari script and English next to each other on same page. Both scripts are being hyphenated based on the custom English_US.pattern file.

Image-2 contains Telugu script being hyphenated based on the same custom English_US.pattern file.

Both screenshots taken from Kobo Libra 2 running KOReader 2024.07. And to the best of my understanding, everything is rendering perfectly well. It was a dream come true when I got a Kobo and got KOReader on it.

![Image-1](https://github.com/user-attachments/assets/333e9668-1875-44e4-a103-e8fc37d5273e)

![Image-2](https://github.com/user-attachments/assets/060c134c-99fc-4ec0-af91-0814a0c8bc39)

**Reply by poire-z** [reply]
Thank you :)

> And to the best of my understanding, everything is rendering perfectly well. It was a dream come true when I got a Kobo and got KOReader on it.

And these scripts do use our western hyphenation sign (minus) for hyphenation, even in real quality paper books ? There's no real other "symbol" in these scripts for hyphenation ? (I guess it may have been taken from English habits.)

**Reply by Frenzie** [reply]
https://www.w3.org/International/ilreq/devanagari/#h_line_breaking doesn't seem to indicate anything I would classify as special.

**Reply by PrasiNV** [reply]
> And these scripts do use our western hyphenation sign (minus) for hyphenation, even in real quality paper books ? There's no real other "symbol" in these scripts for hyphenation ? (I guess it may have been taken from English habits.)

Yes. In print books as well (as old as early last century that I get to see in archive.org), the western hyphenation sign (minus) is used for hyphenation. I never came across a print book which has any other sign. As the Indian scripts (Devanagari, Telugu, Kannada that I know) themselves do not have a minus sign to indicate anything, I think there is no issue to insert minus sign. Hence I said, all is ok "to the best of my understanding".
